File: frontend/src/app/models/ingredient.ts
export interface Ingredient {
    selected: boolean; // generated & used by frontend only
    isCustom: boolean // generated & used by frontend only
    name: string;
    unit: string;
    defaultQuantity: number;
    autoAdd: boolean;
}

interface IngredientQuantity {
    unit: string;
    value: number;
}

export class IngredientForRecipe {
    isCustom: boolean // generated & used by frontend only
    name!: string;
    quantity!: IngredientQuantity;
    withQuantity!: boolean;
    constructor(ingredient: Ingredient) {
        this.quantity       = {unit: ingredient.unit, value: ingredient.defaultQuantity};
        this.name           = ingredient.name;
        this.isCustom       = ingredient.isCustom;
        this.withQuantity   = false;
    }
}

File: backend/ingredient.py
import dataclasses


@dataclasses.dataclass
class IngredientQuantity:
    unit: str
    value: float


@dataclasses.dataclass
class Ingredient:
    name: str
    quantity: IngredientQuantity

    @staticmethod
    def from_dict(data: dict) -> 'Ingredient':
        return Ingredient(
            name=data['name'],
            quantity=IngredientQuantity(**data['quantity'])
        )
